datumaro/plugins/splitter.py

```diff
@@ -0,0 +1,480 @@
+                if ann.type == AnnotationType.label:
+                    labels.append(ann)
+            assert len(labels) == 1, \
+                "Expected exact one label for a DatasetItem"
+            ann = labels[0]
+            if not hasattr(ann, 'label') or ann.label is None:
+                label = None
+            else:
+                label = str(ann.label)
+            if label not in by_labels:
+                by_labels[label] = []
+            by_labels[label].append((idx, ann))
+
+        by_splits = dict()
+        for subset in self._subsets:
+            by_splits[subset] = []
+
+        ## 2. group by attributes
+        required = self._get_required(sratio)
+        for label, items in by_labels.items():
+            np.random.shuffle(items)
+            by_attributes = self._group_by_attributes(items)
+            for attributes, indice in by_attributes.items():
+                gname = 'label: {}, attributes: {}'.format(label, attributes)
+                splits = self._split_indice(indice, gname, sratio, required)
+                for subset, split in zip(subsets, splits):
+                    if len(split) > 0:
+                        by_splits[subset].extend(split)
+
+        for subset in self._subsets:
+            self._parts.append((set(by_splits[subset]), subset))
+
+class SplitforMatchingReID(TaskSpecificSplitter):
+    """
+    Splits dataset for matching, especially re-id task.|n
+    First, splits dataset into 'train+val' and 'test' sets by person id.|n
+    Note that this splitting is not by DatasetItem. |n
+    Then, tags 'test' into 'gallery'/'query' in class-wise random manner.|n
+    Then, splits 'train+val' into 'train'/'val' sets in the same way.|n
+    Therefore, the final subsets would be 'train', 'val', 'test'. |n
+    And 'gallery', 'query' are tagged using anntoation group.|n
+    You can get the 'gallery' and 'query' subsets using 'get_subset_by_group'.|n
+    Notes:|n
+    - Single label is expected for each DatasetItem.|n
+    - Each label is expected to have attribute representing the person id. |n
+    """
+    _group_map = dict()
+
+    def __init__(self, dataset, id_splits, test_splits,
+                 attr_pid="PID", seed=None):
+        """
+        Parameters
+        ----------
+        dataset : Dataset
+        id_splits : list
+            A list of (subset(str), ratio(float))
+            Subset is expected to be one of ["train", "val", "test"].
+            The sum of ratios is expected to be 1.
+        test_splits : list
+            A list of (subset(str), ratio(float))
+            Subset is expected to be one of ["gallery", "query"].
+            The sum of ratios is expected to be 1.
+        attr_pid: str
+            attribute name representing the person id. (default: PID)
+        seed : int, optional
+        """
+        super().__init__(dataset, seed)
+
+        id_subsets, id_ratio = self._validate_splits(id_splits)
+
+        groups = set()
```

Comment:
Let's defer these calculations to the first use of `__iter__`.

Reply:
I don't understand what you mean.. Can you explain more about this?

Reply:
Currently, there is a lot of computations going on in the construction time, including input dataset traversal etc. I suggest moving these computations to the first use - in `__iter__`. This can be done by saving parameters in constructor and introducing a flag variable like `initialized` or something like this. Then, when splits are actually required, compute them.

Input dataset traversal should be considered heavy operation, so let's try to avoid it, when not necessary.

Reply:
I understand now. Thank you, I'll apply your suggestion.

Reply:
If we apply this lazy initialization, some error checking would be delayed, too. 
It's because some errors can be noticed while iterating over the dataset.
So even if XXXSplit is constructed successfully, it can give an error later. 
I don't know this is good for users. How do you think?

Reply:
I think we only need to validate input parameters in constructor. If their application to this specific input dataset produces an error - it looks ok for me to know about this when I run the operation.

Reply:
Thank you for your comment. I'll change the code according to your comment.